pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -0,0 +1,156 @@
+type v1beta1PolicyApplier struct {
+	jwtPolicies []*model.Config
+	// TODO: add mTLS configs.
+	// TODO: add v1alpha1 fallback configs.
+
+	// processedJwtRules is the consolidate JWT rules from all jwtPolicies.
+	processedJwtRules []*v1beta1.JWT
+}
+
+func (a *v1beta1PolicyApplier) JwtFilter(isXDSMarshalingToAnyEnabled bool) *http_conn.HttpFilter {
+	var filterConfigProto *envoy_jwt.JwtAuthentication
+	filterConfigProto = convertToEnvoyJwtConfig(a.processedJwtRules)
+
+	if filterConfigProto == nil {
+		return nil
+	}
+	out := &http_conn.HttpFilter{
+		Name: authn_model.EnvoyJwtFilterName,
+	}
+	if isXDSMarshalingToAnyEnabled {
+		out.ConfigType = &http_conn.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(filterConfigProto)}
+	} else {
+		out.ConfigType = &http_conn.HttpFilter_Config{Config: util.MessageToStruct(filterConfigProto)}
+	}
+	return out
+}
+
+func (a *v1beta1PolicyApplier) AuthNFilter(proxyType model.NodeType, isXDSMarshalingToAnyEnabled bool) *http_conn.HttpFilter {
+	// TODO(diemtvu) implement this.
+	log.Errorf("AuthNFilter(%v, %v) is not yet implemented", proxyType, isXDSMarshalingToAnyEnabled)
+	return nil
+}
+
+func (a *v1beta1PolicyApplier) InboundFilterChain(sdsUdsPath string, meta *model.NodeMetadata) []plugin.FilterChain {
+	// TODO(diemtvu) implement this.
+	log.Errorf("InboundFilterChain is not yet implemented")
+	return nil
+}
+
+// NewPolicyApplier returns new applier for v1beta1 authentication policies.
+func NewPolicyApplier(jwtPolicies []*model.Config) authn.PolicyApplier {
+	a := &v1beta1PolicyApplier{
+		jwtPolicies: jwtPolicies,
+	}
+	a.aggregateJWTRules()
+
+	return a
+}
+
+func (a *v1beta1PolicyApplier) aggregateJWTRules() {
+	a.processedJwtRules = []*v1beta1.JWT{}
+	// TODO(diemtvu) should we need to deduplicate JWT with the same issuer.
+	for idx := range a.jwtPolicies {
+		spec := a.jwtPolicies[idx].Spec.(*v1beta1.RequestAuthentication)
+		for _, rule := range spec.JwtRules {
+			a.processedJwtRules = append(a.processedJwtRules, rule)
+		}
+	}
+}
+
+func convertToEnvoyJwtConfig(jwtRules []*v1beta1.JWT) *envoy_jwt.JwtAuthentication {
+	if len(jwtRules) == 0 {
+		return nil
+	}
+	providers := map[string]*envoy_jwt.JwtProvider{}
+	for i, jwtRule := range jwtRules {
+		// TODO(diemtvu): set forward based on input spec after https://github.com/istio/api/pull/1172
+		provider := &envoy_jwt.JwtProvider{
+			Issuer:            jwtRule.Issuer,
+			Audiences:         jwtRule.Audiences,
+			Forward:           true,
+			PayloadInMetadata: jwtRule.Issuer,
+		}
+
+		for _, location := range jwtRule.FromHeaders {
+			provider.FromHeaders = append(provider.FromHeaders, &envoy_jwt.JwtHeader{
+				Name:        location.Name,
+				ValuePrefix: location.Prefix,
+			})
+		}
+		provider.FromParams = jwtRule.FromParams
+
+		jwtPubKey := jwtRule.Jwks
+		if jwtPubKey == "" {
+			var err error
+			jwtPubKey, err = model.JwtKeyResolver.GetPublicKey(jwtRule.JwksUri)
+			if err != nil {
```

Comment:
this function does not return error, and this error does nothing other than logging to error, in case of pk fetching error?

Is that desired and what we're doing now?

Reply:
Yes it is. At this point, not much pilot can do. Setting a jwt filter with empty public key will guarantee the validation will fail, and thus reject all requests. Mesh will be broken, but better than breach.

Btw, in this new implementation, we will have the error message from JWT filter return together with the 401, so operators can tell what happen.

Reply:
possible to add a test case to simulate error handling with empty filing then?

Reply:
It was alread there, under `TestConvertToEnvoyJwtConfig`. Just added another one for `TestJwtFilter` as well, just to be thorough.